validator/src/service/machine.ts

```diff
@@ -0,0 +1,499 @@
+	| {
+			id: "waiting_for_rollover";
+	  }
+	| {
+			id: "collecting_commitments";
+			groupId: GroupId;
+			nextEpoch: bigint;
+	  }
+	| {
+			id: "collecting_shares";
+			groupId: GroupId;
+			nextEpoch: bigint;
+			lastParticipant?: ParticipantId;
+	  }
+	| {
+			id: "request_rollover_data";
+			groupId: GroupId;
+			nextEpoch: bigint;
+	  }
+	| {
+			id: "sign_rollover_msg";
+			groupId: GroupId;
+			nextEpoch: bigint;
+			msg: Hex;
+	  };
+
+type SigningState =
+	| {
+			id: "collect_nonce_commitments";
+	  }
+	| {
+			id: "collect_signing_shares";
+			lastSigner?: ParticipantId;
+	  }
+	| {
+			id: "signed";
+	  };
+
+export class ShieldnetStateMachine {
+	#participants: Participant[];
+	#protocol: ShieldnetProtocol;
+	#verificationEngine: VerificationEngine;
+	#keyGenClient: KeyGenClient;
+	#signingClient: SigningClient;
+	#activeEpoch;
+	#genesisGroupId?: GroupId;
+	#logger?: (msg: unknown) => void;
+	#blocksPerEpoch: bigint;
+
+	#lastProcessedBlock = 0n;
+	#lastProcessedIndex = 0;
+	#stagedEpoch = 0n;
+
+	#keyGenState: KeyGenState = { id: "waiting_for_rollover" };
+	#signingState = new Map<SignatureId, SigningState>();
+
+	#epochGroups = new Map<bigint, GroupId>();
+	#groupSequence = new Map<GroupId, bigint>();
+	#groupPendingNonces = new Set<GroupId>();
+
+	constructor({
+		participants,
+		protocol,
+		keyGenClient,
+		signingClient,
+		verificationEngine,
+		initialEpoch,
+		logger,
+		blocksPerEpoch,
+	}: {
+		participants: Participant[];
+		protocol: ShieldnetProtocol;
+		keyGenClient: KeyGenClient;
+		signingClient: SigningClient;
+		verificationEngine: VerificationEngine;
+		initialEpoch?: bigint;
+		logger?: (msg: unknown) => void;
+		blocksPerEpoch?: bigint;
+	}) {
+		this.#participants = participants;
+		this.#protocol = protocol;
+		this.#keyGenClient = keyGenClient;
+		this.#signingClient = signingClient;
+		this.#verificationEngine = verificationEngine;
+		this.#activeEpoch = initialEpoch ?? 0n;
+		this.#logger = logger;
+		this.#blocksPerEpoch = blocksPerEpoch ?? BLOCKS_PER_EPOCH;
+	}
+
+	async progressToBlock(block: bigint) {
+		if (block < this.#lastProcessedBlock) {
+			throw Error(`There is no going back!`);
+		}
+		// Check if we are already up to date
+		if (block === this.#lastProcessedBlock) return;
+		this.#lastProcessedBlock = block;
+
+		await this.checkGenesis();
+		await this.checkEpochRollover(block);
+		await this.checkAvailableNonces();
+
+		// TODO:
+		// Check keyGen timeouts
+		// Check signing timeouts
+	}
+
+	async processBlockEvent(
+		block: bigint,
+		index: number,
+		eventName: string,
+		eventArgs: unknown,
+	) {
+		if (
+			block < this.#lastProcessedBlock ||
+			(block === this.#lastProcessedBlock && index <= this.#lastProcessedIndex)
+		) {
+			throw Error(
+				`Invalid block number (${block}) and index ${index} (currentyl at block ${this.#lastProcessedBlock} and index ${this.#lastProcessedIndex})`,
+			);
+		}
+		this.#lastProcessedIndex = index;
+		await this.progressToBlock(block);
```

Comment:
Theoretically, the block is processed _before_ the event index. I don't think it matters for this class though.